src/wayland/shell/xdg/mod.rs

```diff
@@ -108,36 +113,270 @@ mod xdg_handlers;
+
+                /// Holds the state if the surface has sent the initial
+                /// configure event to the client. It is expected that
+                /// during the first commit a initial
+                /// configure event is sent to the client
+                pub initial_configure_sent: bool,
+
+                /// Holds the configures the server has sent out
+                /// to the client waiting to be acknowledged by
+                /// the client. All pending configures that are older
+                /// than the acknowledged one will be discarded during
+                /// processing xdg_surface.ack_configure.
+                pending_configures: Vec<$configure>,
+
+                $(
+                    $(#[$field_attr])*
+                    $vis $field: $type,
+                )*
+        }
+
+        impl XdgSurface for $element {
+            fn set_window_geometry(&mut self, geometry: Rectangle) {
+                self.next_geometry = Some(geometry);
+            }
+
+            fn ack_configure(&mut self, serial: Serial) -> Result<Configure, ConfigureError> {
+                let configure = match self
+                    .pending_configures
+                    .iter()
+                    .find(|configure| configure.serial == serial)
+                {
+                    Some(configure) => (*configure).clone(),
+                    None => {
+                        return Err(ConfigureError::SerialNotFound(serial));
+                    }
+                };
+
+                // Role specific ack_configure
+                let role_ack_configure: &dyn Fn(&mut Self, $configure) = &$role_ack_configure;
+                role_ack_configure(self, configure.clone());
+
+                // Set the xdg_surface to configured
+                self.configured = true;
+
+                // Save the last configure serial as a reference
+                self.configure_serial = Some(Serial::from(serial));
+
+                // Clean old configures
+                self.pending_configures.retain(|c| c.serial > serial);
+
+                Ok(configure.into())
+            }
+        }
+
+        impl Default for $element {
+            fn default() -> Self {
+                Self {
+                    geometry: None,
+                    next_geometry: None,
+                    configured: false,
+                    configure_serial: None,
+                    configure_next_serial: Serial::from(0),
+                    pending_configures: Vec::new(),
+                    initial_configure_sent: false,
+
+                    $(
+                        $field: Default::default(),
+                    )*
+                }
+            }
+        }
+    };
+}
+
+trait CompositorTokenExt {
+    fn with_xdg_role<F, T>(&self, surface: &wl_surface::WlSurface, f: F) -> Result<T, WrongRole>
+    where
+        F: FnOnce(&mut dyn XdgSurface) -> T;
+}
+
+impl<R> CompositorTokenExt for CompositorToken<R>
```

Comment:
`CompositorToken` is defined in the same crate, we don't need an extension trait to add new methods to it.

Reply:
That would add a reference to the `XdgSurface` trait inside the `compositor` mod which i assumed would be against the modular design of smithay. I could just use a helper function here, i just did not want to repeat the checks for both roles
all over the xdg implementation.

Reply:
The impl would not need to be inside the compositor mod, you can add methods from here, unless that has recently been changed in rustc?

Reply:
Right, sorry for the fuzz, i was mistakenly thinking the token was a trait 😅

Reply:
I have removed the trait and implemented it directly.